File: mods/zdtd_tradefeed/zdtd_tradefeed.c
// zdtd_tradefeed - a trader-event observer plugin (AGENTS.md rule 29,
// Wasm-first). Reference module for the on_trader_event hook: it logs trader
// window opens, buys and sells (kind 0/1/2). Announcements live in the
// module; the server only fires the event.
//
// Build (clang, committed as mods/zdtd_tradefeed/zdtd_tradefeed.wasm):
//   clang --target=wasm32 -nostdlib -O2 -Wl,--no-entry -Wl,--export-all \
//     -o mods/zdtd_tradefeed/zdtd_tradefeed.wasm mods/zdtd_tradefeed/zdtd_tradefeed.c

__attribute__((import_module("zdtd"), import_name("log")))
extern void zdtd_log(int level, int ptr, int len);

#define OUT_CAP 160
static char out[OUT_CAP];
static int out_n;

static void e(const char *s) {
  while (*s && out_n < OUT_CAP - 1) out[out_n++] = *s++;
}
static void e_int(long long v) {
  char t[24];
  int n = 0;
  if (v == 0) {
    t[n++] = '0';
  } else {
    if (v < 0) { if (out_n < OUT_CAP - 1) out[out_n++] = '-'; v = -v; }
    while (v > 0 && n < 23) { t[n++] = (char)('0' + (int)(v % 10)); v /= 10; }
  }
  int i;
  for (i = n - 1; i >= 0; --i) {
    if (out_n < OUT_CAP - 1) out[out_n++] = t[i];
  }
}
static void flush(int level) {
  if (out_n > 0) {
    zdtd_log(level, (int)(long)&out[0], out_n);
    out_n = 0;
  }
}

void on_enable(void) {
  out_n = 0;
  e("zdtd_tradefeed v1.0 enabled (trader events)");
  flush(1);
}

void on_shutdown(void) {
  out_n = 0;
  e("zdtd_tradefeed shutdown");
  flush(1);
}

void on_trader_event(int player, int trader, int kind) {
  out_n = 0;
  e("trader event: player=");
  e_int(player);
  e(" trader=");
  e_int(trader);
  e(" kind=");
  switch (kind) {
    case 0: e("open"); break;
    case 1: e("buy"); break;
    case 2: e("sell"); break;
    default: e_int(kind); break;
  }
  flush(1);
}

File: mods/zdtd_tradefeed/zdtd_tradefeed.zig
// zdtd_tradefeed — a trader-event observer plugin (AGENTS.md rule 29,
// Wasm-first). Reference module for trader/vehicle announcements: observes
// on_trader_event (kind 0 open / 1 buy / 2 sell) and logs every trade window
// event, always keeping the stock outcome.
//
// Build (zig): see mods/BUILDING.md. Committed as zdtd_tradefeed.wasm.

const common = @import("plugin_common");

var out: common.Buf = .{};

export fn on_enable() void {
    out.reset();
    out.put("zdtd_tradefeed v1.0 enabled (trader events)");
    out.logLine(1);
}

export fn on_shutdown() void {
    out.reset();
    out.put("zdtd_tradefeed shutdown");
    out.logLine(1);
}

export fn on_trader_event(player: i32, trader: i32, kind: i32) void {
    out.reset();
    out.put("trader event: player=");
    out.putInt(player);
    out.put(" trader=");
    out.putInt(trader);
    out.put(" kind=");
    switch (kind) {
        0 => out.put("open"),
        1 => out.put("buy"),
        2 => out.put("sell"),
        else => out.putInt(kind),
    }
    out.logLine(1);
}

comptime {
    // Declarative dependency check (paper: reactive coeffects): the host
    // rejects the module at load if any listed capability is missing.
    common.exportRequires("on_trader_event,log");
}
